### DISCUSSION TITLE
ScriptReference/EditorApplication.ExecuteMenuItem

### DISCUSSION THREAD

**Discussion by docscomments** [Developer-Notes-Extension/Unity-Notes#214] [Notes]
# ScriptReference/EditorApplication.ExecuteMenuItem

https://docs.unity3d.com/ScriptReference/EditorApplication.ExecuteMenuItem.html

**Comment by mvi** [comment]
Note you can find a list of all menu items by using the undocumented `EditorGUIUtility.SerializeMainMenuToString()` method.  
To generate complete paths, use something like the following (tested on Unity 2018.3)  
  

```csharp
static int CalculateIndent(string source)  
{  
    int spaceCount = 0;  
    for (int i = 0; i < source.Length; i++)  
    {  
        if (source[i] == ' ')  
        {  
            spaceCount++;  
        }  
        else  
        {  
            break;  
        }  
    }  
  
    return spaceCount / 4;  
}  
  
public static List<string> ParseMenuItems()  
{  
    List<string> menuItemMappings = new List<string>();  
    string mainMenuString = EditorGUIUtility.SerializeMainMenuToString();  
    string[] split = mainMenuString.Split('\n');  
    int lastIndent = -1;  
  
    List<string> stack = new List<string>();  
  
    foreach (string line in split)  
    {  
        if (line.Trim().Length == 0) continue;  
  
        int newIndent = CalculateIndent(line);  
        for (int i = lastIndent; i < newIndent; i++)  
        {  
            stack.Add("");  
        }  
        for (int i = lastIndent; i > newIndent; i--)  
        {  
            stack.RemoveAt(stack.Count - 1);  
        }  
  
        stack[stack.Count - 1] = line.Trim();  
  
        string path = string.Join("/", stack.ToArray());  
        menuItemMappings.Add(path);  
  
        lastIndent = newIndent;  
    }  
  
    return menuItemMappings;  
}
```

**Comment by mvi** [comment]
Note that from Unity 2018.2 almost all built in Window paths change, for example Window/Inspector becomes Window/General/Inspector. As such you may need to use #if UNITY_2018_2_OR_NEWER to branch the paths in your code if you need to support older versions of Unity.  
A handy use of ExecuteMenuItem is to combine it with a selection change. For example if you have a config scriptable object that you want to link to from a custom editor window you could write something like:  
  

```csharp
Selection.activeObject = KeyMappings.Instance;  
EditorApplication.ExecuteMenuItem("Window/General/Inspector");
```
  
This sets the scriptable object (or any object) as the selection then will either focus an inactive Inspector tab or create a new one if there isn't one already.
